packages/react/src/useMarkprompt.tsx

```diff
@@ -0,0 +1,130 @@
+import {
+  DEFAULT_MODEL,
+  DEFAULT_I_DONT_KNOW_MESSAGE,
+  MARKPROMPT_COMPLETIONS_URL,
+  submitPrompt,
+  DEFAULT_LOADING_HEADING,
+  DEFAULT_REFERENCES_HEADING,
+  DEFAULT_PROMPT_TEMPLATE,
+  DEFAULT_TOP_P,
+  DEFAULT_TEMPERATURE,
+  DEFAULT_FREQUENCY_PENALTY,
+  DEFAULT_PRESENCE_PENALTY,
+  DEFAULT_MAX_TOKENS,
+} from '@markprompt/core';
+import type { Options } from '@markprompt/core';
+import { useCallback, useMemo, useRef, useState } from 'react';
+
+export function useMarkprompt({
+  projectKey,
+  completionsUrl = MARKPROMPT_COMPLETIONS_URL,
+  iDontKnowMessage = DEFAULT_I_DONT_KNOW_MESSAGE,
+  referencesHeading = DEFAULT_REFERENCES_HEADING,
+  loadingHeading = DEFAULT_LOADING_HEADING,
+  includeBranding = true,
+  model = DEFAULT_MODEL,
+  promptTemplate = DEFAULT_PROMPT_TEMPLATE,
+  temperature = DEFAULT_TEMPERATURE,
+  topP = DEFAULT_TOP_P,
+  frequencyPenalty = DEFAULT_FREQUENCY_PENALTY,
+  presencePenalty = DEFAULT_PRESENCE_PENALTY,
+  maxTokens = DEFAULT_MAX_TOKENS,
+}: { projectKey: string } & Options) {
+  if (!projectKey) {
+    throw new Error(
+      'Markprompt: a project key is required. Make sure to pass the projectKey prop to Markprompt.Root.',
+    );
+  }
+
+  const [answer, setAnswer] = useState('');
+  const [references, setReferences] = useState<string[]>([]);
+  const [loading, setLoading] = useState(false);
+  const [prompt, setPrompt] = useState<string>('');
+
+  const controller = useRef(new AbortController());
+
+  //   useEffect(() => {
+  //     console.log('aborting');
+  //     const c = controller.current;
+  //     return () => c.abort();
+  //   }, []);
+
+  const handlePromptChange = useCallback(
+    (event: React.ChangeEvent<HTMLInputElement>) => {
+      setPrompt(event.target.value);
+    },
+    [],
+  );
+
+  const handleSubmit: React.FormEventHandler<HTMLFormElement> = useCallback(
+    async (event) => {
+      event.preventDefault();
+
+      if (!prompt || prompt === '') {
+        return;
+      }
+
+      setAnswer('');
+      setReferences([]);
+      setLoading(true);
+
+      await submitPrompt(
+        prompt,
+        projectKey,
+        (chunk) => {
+          setAnswer((prev) => prev + chunk);
+          return true;
+        },
+        (refs) => setReferences(refs),
+        (error) => {
+          console.error(error);
+        },
+        {
+          completionsUrl,
+          iDontKnowMessage,
+          referencesHeading,
+          loadingHeading,
+          includeBranding,
+          model,
+          promptTemplate,
+          temperature,
+          topP,
+          frequencyPenalty,
+          presencePenalty,
+          maxTokens,
+          signal: controller.current.signal,
```

Comment:
This controller is never aborted.

https://github.com/motifland/markprompt-js/pull/5/files#diff-ed3c66975aa962b6eb6cf98d8a89085b627bbff83459263ccd36415033866284R83-R118 changes this to a `useEffect()`, which makes it clear when to create a controller, and when to abort it. With a `useCallback()` this is not as clear.

I think something like this could work:

```js
const controllerRef = useRef<AbortController>()

const submit = useCallback((event) => {
  controllerRef.current?.abort();

  const controller = new AbortController()
  controllerRef.current = controller

  const promise = submitPrompt(
    // …
    {
      // …
      signal: controller.signal
    }
  )
  promise.finally(() => {
    if(controllerRef.current === controller) {
      controllerRef.current = undefined
    }
  })
  return promise
}, [
  // …
])
```

Reply:
Yes! That seems like a good approach. I'll see if I can also add an effect hook that has an unmount callback for cancelling the running fetch on unmount, if any.

Reply:
Good catch on the unmount!

Reply:
I'm also now exposing an `abort` callback in the return value of the hook. We consume that for instance in the dialog's close button click handler.